cwms-data-api/src/test/java/fixtures/KeyCloakExtension.java

```diff
@@ -0,0 +1,155 @@
+    private static String issuer;
+    private static String codeUrl;
+    private static String tokenUrl;
+    
+
+    
+    static void setup() throws IOException {
+        
+        File realm = new File("../compose_files/keycloak/realm.json").getAbsoluteFile();
+        String realmJson = null;
+        try (FileInputStream is = new FileInputStream(realm)) {
+            realmJson = IOUtils.toString(is, Charset.forName("UTF-8"));
+        }
+        kcc.withCopyToContainer(Transferable.of(realmJson), "/opt/keycloak/data/import/realm.json");
+        kcc.setWaitStrategy(
+            new LogMessageWaitStrategy().withRegEx("^.*Listening on:.*$")
+                                        .withTimes(1)
+                                        .withStartupTimeout(Duration.ofMinutes(5))
+        );
+        kcc.start();
+        authUrl = "http://"+kcc.getHost()+":"+kcc.getMappedPort(8080);
+        
+        // verify we can get the wellknown config
+        Response response = 
+            given()
+                .log().ifValidationFails(LogDetail.ALL,true)
+                .baseUri(authUrl)
+            .when()
+                .get(WELL_KNOWN);
+        response
+            .then()
+            .statusCode(is(HttpCode.OK.getStatus()))
+            ;
+        
+        JsonNode oidcConfig = mapper.readTree(response.asPrettyString());
+        issuer = oidcConfig.get("issuer").asText();
+        codeUrl = oidcConfig.get("authorization_endpoint").asText();
+        tokenUrl = oidcConfig.get("token_endpoint").asText();
+        logger.fine(response::asPrettyString);
+    }
+
+    @Override
+    public void beforeAll(ExtensionContext context) throws Exception {
+        if (!kcc.isRunning()) {
+            setup();
+        }
+    }
+
+    public static String getAuthUrl() {
+        return authUrl;
+    }
+
+    public static String getOidcWellKnown() {
+        return authUrl+"/"+WELL_KNOWN;
+    }
+
+    public static String getIssuer() {
+        return issuer;
+    }
+
+    public static String getCodeUrl() {
+        return codeUrl;
+    }
+
+    public static String getTokenUrl() {
+        return tokenUrl;
+    }
+    
+    /**
+     * Retrieve the Access token for the user
+     * @param username
+     * @param password
+     * @return Access token only
+     */
+    public static Optional<String> tokenForUser(String username, String password) {
+        try {
+            Response response = 
+                given()
+                    .log().ifValidationFails(LogDetail.ALL,true)
+                    .contentType(ContentType.URLENC)
+                    .formParam("client_id","cwms")
+                    .formParam("grant_type","password")
+                    .formParam("client_secret","")
+                    .formParam("scope","openid profile email")
+                    .formParam("username",username)
+                    .formParam("password",password)
+                    .formParam("response_type","token")
```

Comment:
The keys here could probably be constants too

Reply:
Probably, but at this time this is the only place they're used. If we end up duplicating this, say for testing AuthCode + PKCE, that's probably the better time to do that. (And yes, I'm kinda just being a bit lazy.)